### PR TITLE
feat: add MercadoPago payment app

### PR DESCRIPTION
**Pull request by ubinatus**
## What does this PR do?

1. Adds MercadoPago as a new payment app for Cal.com,
2. Adds a webhook endpoint to listen for payments created with MercadoPago.

#### About MercadoPago
Mercado Pago is a leading digital payment platform in Latin America. Currently, it's available in seven countries: Argentina, Brazil, Chile, Colombia, Mexico, Uruguay, and Peru. It's widely used by businesses and consumers in all the above countries, and it surpassed $20 billion payment volume in 2019.

Checkout the [Loom Video Demo](https://www.loom.com/share/5746c7952fb444a9aec3caeb55f7191c?sid=0d9b74a5-d376-4be5-b750-b64d4bc168f0) for this PR.

## Requirement/Documentation

- If there is a requirement document, please, share it here.
 No
- If there is ab UI/UX design document, please, share it here.
 No

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How should this be tested?

- Are there environment variables that should be set?
-- Yes, we need the following environment variables: `MERCADOPAGO_CLIENT_ID`, `MERCADOPAGO_CLIENT_SECRET`,  `NEXT_PUBLIC_MERCADOPAGO_PUBLIC_KEY`. See the guide after this bullet points to learn how to retrieve the configs from the App in MercadoPago to Cal.com enviornment variables.
- What are the minimal test data to have?
-- _MercadoPago requirements_: 
--- 1. Signup to MercadoPago [here](https://www.mercadopago.com).
--- 2. Create MercadoPago application in the developers portal [here](https://www.mercadopago.com.pe/developers/panel/app).
--- 3. Retrieve the Client ID on the applications page.
--- 4. Retrieve the Secret on the test credentials section (sidebar).
--- 5. Retrieve the Public Key on the test credentials section (sidebar)
--- 6. Set the Webhook (sidebar). Should point to `YOUR_URL/api/integrations/mercadopago/webhook` and should select "Pagos" (Payments) as the only event needed.
--- 7. Set the Redirect Url on the the application page. Should point to `YOUR_URL/api/integrations/mercadopago/callback`
See the guide after this bullet points to learn how to setup an App in MercadoPago.
-- _Cal.com requirements:_
--- 1. Install MercadoPago payment app.
--- 2. Add the MercadoPago payment information in an EventType
- What is expected (happy path) to have (input and output)?
-- Being able to install MercadoPago payment app in Cal.com.
-- Being able to book and pay an event with MercadoPago in Cal.com.
-- Being able to cancel an event and refund with MercadoPago in Cal.com.
- Any other important info that could help to test that PR.
-- See the following guide.

In order to test the MercadoPago integration:
1. Create a MercadoPago account [here](https://www.mercadopago.com/developers/panel/app). 
2. Go to the [developers site](https://www.mercadopago.com/developers/panel/app) and set up an application. Here you should add the relevant environment variables to the `.env` file. You'll need the Application ID of MercadoPago (`MERCADOPAGO_CLIENT_ID`), the Public Key of MercadoPago (`NEXT_PUBLIC_MERCADOPAGO_PUBLIC_KEY`), the Access Token of MercadoPago (`MERCADOPAGO_CLIENT_SECRET`). After that you will to set up the Webhook URL to point to the `https://yesyesyes.ngrok.app/api/integrations/mercadopago/webhook`. Finally, you should edit the MercadoPago application and add the RedirectURL to point to `https://yesyesyes.ngrok.app/api/integrations/mercadopago/callback`. I've made a [Loom Video](https://www.loom.com/share/9b7d8a04a476496b9b2b6fd0cdf508fb?sid=59614e50-4455-451e-b342-652103592745) to quickly explain this process.
3. Once you had set up the configs in the `.env` file you should run `yarn dx` so that it can seed again the app credentials to the database.
4. Install MercadoPago payment app to your Cal.com account.
5. Create an EventType and add the MercadoPago app. Setup the price and currency. The currency should be the same local currency (or USD) as the business set up in MercadoPago. Since there is no way to retrieve this from the API, the user should manually set the currency.
6. In an incognito window, schedule an Event with the EventType that has the MercadoPago integration. 
7. After confirming the booking you will see the Payment page with the MercadoPago payment button. The currency and price should be reflected here.
9. The pay button will redirect you to MercadoPago account so that you can proceed with the payment. You can used the [test credit cards](https://www.mercadopago.com.pe/developers/es/docs/checkout-pro/additional-content/your-integrations/test/cards) and enter "12345678" for the ID number.
10. After the payment is done you should be redirected to the Booking page, which will indicate that the payment was made.
11. You can also cancel the event and will see that the Payment has been refunded by MercadoPago.

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: packages/app-store/mercadopago/config.json
Comment (ubinatus) [inline packages/app-store/mercadopago/config.json:6]: TODO: add the MercadoPago url `https://mercadopago.com`

File: packages/app-store/mercadopago/lib/PaymentService.ts
Comment (ubinatus) [inline packages/app-store/mercadopago/lib/PaymentService.ts:86]: TODO: remove `external_reference` param as it's not passed.

File: packages/app-store/mercadopago/lib/PaymentService.ts
Comment (ubinatus) [inline packages/app-store/mercadopago/lib/PaymentService.ts:105]: Here we don't have access to the MercadoPago (MP) `Payment` yet. The preference lets the user create a payment when redirected to the MP checkout. It will forward the metadata and the Cal.com `uid`.

File: packages/app-store/mercadopago/lib/constants.ts
Comment (ubinatus) [inline packages/app-store/mercadopago/lib/constants.ts:10]: There is no auth/capture when using MercadoPago preferences.

File: packages/app-store/mercadopago/lib/MercadoPago.ts
Comment (ubinatus) [inline packages/app-store/mercadopago/lib/MercadoPago.ts:106]: Here, we add the `expires_at` value since it is not present on the OAuth response.

File: packages/features/ee/payments/api/mercadopago-webhook.ts
Comment (ubinatus) [inline packages/features/ee/payments/api/mercadopago-webhook.ts:43]: Unlike Stripe and Paypal webhooks, here the event itself is not going to help us retrieve the payment/booking from Cal.com database. We will need to query the full payment from MercadoPago so that we can know which Cal.com payment/booking is associated with.

File: packages/features/ee/payments/api/mercadopago-webhook.ts
Comment (ubinatus) [inline packages/features/ee/payments/api/mercadopago-webhook.ts:184]: Unlike Stripe and Paypal, we here now store the `MercadoPago Payment` on the Cal.com Payment.

File: packages/prisma/seed-app-store.ts
Comment (ubinatus) [inline packages/prisma/seed-app-store.ts:374]: Seeding the DB with MP credentials

File: packages/trpc/server/routers/viewer/eventTypes/update.handler.ts
Comment (ubinatus) [inline packages/trpc/server/routers/viewer/eventTypes/update.handler.ts:274]: Addresses @zomars comment: https://github.com/calcom/cal.com/pull/11181/files#r1317740686, which might solve #11181.

**Comment by alannnc** [comment]
Loom link is broken, can you re upload it please?

**Comment by zomars** [comment]
Thank you for your contribution 🙏 It seems the loom video isn't public. I can't really see it.

**Comment by ubinatus** [comment]
@alannnc @zomars Apologies for the broken link. Description should with latest Loom url should be updated now. 
There are 2 loom videos mentioned on the PR description:
- [Feature showcase](https://www.loom.com/share/5746c7952fb444a9aec3caeb55f7191c?sid=38169a7b-e22b-45ee-9a8c-360cf2bb730f)
- [MercadoPago setup guide](https://www.loom.com/share/9b7d8a04a476496b9b2b6fd0cdf508fb?sid=2b6fbf63-b0a9-4306-8c22-ce7eadf9fe62)

**Comment by PeerRich** [comment]
high prio becasue its been way overdue cc @alannnc for testing and review

**Comment by zomars** [comment]
> high prio becasue its been way overdue

I don't think that's reason for it being high priority. Getting high demand for the feature would be in this case. Still will get into this ASAP.

**Review by alannnc — COMMENTED** [review]
~~Looks to me that this will only work for self hosters, we need to obtain user MP keys so they can install it on cal.com.~~

Do you think we should be better off using https://www.npmjs.com/package/mercadopago instead of the custom lib?

File: apps/web/pages/booking/[uid].tsx
Comment (alannnc) [inline apps/web/pages/booking/[uid].tsx:93]: This currently has no use.

**Comment by ubinatus** [comment]
> ~Looks to me that this will only work for self hosters, we need to obtain user MP keys so they can install it on cal.com.~
> 
> Do you think we should be better off using https://www.npmjs.com/package/mercadopago instead of the custom lib?

Issue with mercadopago sdk is that it's currently throwing a lot vulnerabilities when running the `audit`. It still uses the deprecated `request` package (see dis [here](https://github.com/mercadopago/sdk-nodejs/discussions/233#discussioncomment-6671776)). Thus, current Cal.com usage for payments won't consume more than 20% of the existing api methods from MP. 

I ended up removing the sdk and just calling the necessary [API](https://www.mercadopago.com.pe/developers/es/reference) endpoints.

**Comment by alannnc** [comment]
@ubinatus Do you think this will work for multi countries? e.g. I setup the main app in an account created on MX but you as a client sign up using you MP account from Peru. Will payments be available in your currency?

I ask this because I seem to not be able to create test users for a different country.

**Comment by ubinatus** [comment]
Hi Alan. Great question. Just researched it on MP discord and _seems_ that a MP application creating in country X can only allow integrations with country X (see dis [here](https://discord.com/channels/1078660903450464286/1149843675145310328)). I'll have to test this to make sure.

If that's the case, then the config for Cal.com will require to create a MP account in each country (Argentina, Brazil, Chile, Colombia, Mexico, Uruguay, and Peru), which will make the config more complex.

**Comment by alannnc** [comment]
> Hi Alan. Great question. Just researched it on MP discord and _seems_ that a MP application created in country X can only allow integrations with country X (see dis [here](https://discord.com/channels/1078660903450464286/1149843675145310328)). I'll have to test this to make sure.
> 
> If that's the case, then the config for Cal.com will require to create a MP account in each country (Argentina, Brazil, Chile, Colombia, Mexico, Uruguay, and Peru), which will make the config more complex.

Yeah let's try it out, I can give you my public url so you can jump to my local and try to connect your MP account from Peru. Or you can give me yours. Also try to look me on cal.com discord for a fast reply. 
username: alan.castro

@ubinatus

**Comment by ubinatus** [comment]
@alannnc Sure thing! Let's connect to solve this. Will ping you there in a while. (Not near my pc)

**Comment by alannnc** [comment]
@ubinatus Thanks for your time it was a good debugging session, we are gonna revaluate what is the best way to handle the changes needed to bring this to cal.com where self host and multi country is supported or don't give that much trouble to all.

**Review by zomars — CHANGES_REQUESTED** [review]
Putting on hold until we decide what path to take for MP. 🙏🏽

**Comment by ubinatus** [comment]
> > ~Looks to me that this will only work for self hosters, we need to obtain user MP keys so they can install it on cal.com.~
> > Do you think we should be better off using https://www.npmjs.com/package/mercadopago instead of the custom lib?
> 
> Issue with mercadopago sdk is that it's currently throwing a lot vulnerabilities when running the `audit`. It still uses the deprecated `request` package (see dis [here](https://github.com/mercadopago/sdk-nodejs/discussions/233#discussioncomment-6671776)). Thus, current Cal.com usage for payments won't consume more than 20% of the existing api methods from MP.
> 
> I ended up removing the sdk and just calling the necessary [API](https://www.mercadopago.com.pe/developers/es/reference) endpoints.

Heads up, last week MercadoPago released their latest type safe sdk, which solves the audit vulnerabilities mentioned.

**Comment by PeerRich** [comment]
> Heads up, last week MercadoPago released their latest type safe sdk, which solves the audit vulnerabilities mentioned.

are we using this here already?

**Comment by alannnc** [comment]
> > Heads up, last week MercadoPago released their latest type safe sdk, which solves the audit vulnerabilities mentioned.
> 
> are we using this here already?

This PR is using plain API calls to MP, my PR was using old sdk version.

**Review by joeauyeung — CHANGES_REQUESTED** [review]
If we're going to merge this I would feel better if we used the SDK provided by Mercado Pago. I'm seeing in `packages/app-store/mercadopago/lib/MercadoPago.ts` we're still using API calls

**Comment by keithwillcode** [comment]
Thanks for the detailed input, @ubinatus. Much appreciated. 

It indeed sounds like a bit of a mess at the moment so will close the PR but as you mentioned it can stay as a reference point.

**Pull request closed without merging**
